tests/robustness/traffic/kubernetes.go

```diff
@@ -102,54 +94,92 @@ func (t kubernetesTraffic) Run(ctx context.Context, c *RecordingClient, limiter
 			}
 			// Avoid multiple failed writes in a row
 			if lastWriteFailed {
-				listCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
-				resp, err := kc.List(listCtx, keyPrefix)
-				cancel()
+				_, err := t.Read(ctx, kc, s, limiter, keyPrefix)
 				if err != nil {
 					continue
 				}
-				s.Reset(resp)
-				limiter.Wait(ctx)
 			}
-			writeCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
-			err := t.Write(writeCtx, kc, ids, s)
-			cancel()
+			err := t.Write(ctx, kc, ids, s, limiter)
 			lastWriteFailed = err != nil
 			if err != nil {
 				continue
 			}
-			limiter.Wait(ctx)
 		}
 	})
 	g.Wait()
 }
 
-func (t kubernetesTraffic) Write(ctx context.Context, kc *kubernetesClient, ids identity.Provider, s *storage) (err error) {
+func (t kubernetesTraffic) Read(ctx context.Context, kc *kubernetesClient, s *storage, limiter *rate.Limiter, keyPrefix string) (rev int64, err error) {
+	limit := int64(t.averageKeyCount)
+	var revision int64 = 0
+	readCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
+	resp, err := kc.List(readCtx, keyPrefix, revision, limit)
+	cancel()
+	if err != nil {
+		return 0, err
+	}
+	limiter.Wait(ctx)
+	hasMore := resp.More
+	revision = resp.Header.Revision
+	kvs := resp.Kvs
+
+	for hasMore {
+		lastKey := string(kvs[len(kvs)-1].Key)
+		readCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
+		resp, err := kc.Range(readCtx, lastKey+"\x00", keyPrefix+"\xff", revision, limit)
```

Comment:
`keyPrefix+"\xff` can be replaced with `clientv3.GetPrefixRangeEnd(prefix)`

Reply:
Done